test/e2e/framework/t.go

```diff
@@ -98,5 +100,24 @@ func (t *T) Wait() {
 	}
 }
 
+func (t *T) flush() {
+	t.T.Helper()
+loop:
+	for {
+		select {
+		case err := <-t.errors:
+			t.T.Helper()
+			t.T.Error(err)
+		default:
+			break loop
+		}
+	}
+}
+
+func (t *T) Failed() bool {
+	t.flush()
```

Comment:
Why is this needed?

Reply:
some go routine does `t.Errorf()` (e.g. the kcp starter in the framework), and calls `wg.Done`. The outer code waits for the wg and then checks `t.Failed()`. Today false is returned, although there is a clear order in time.

Reply:
This code:
```golang
			for _, srv := range servers {
				// binding the server to ctx ensures its lifetime is only
				// as long as the test we are running in this specific case
				if err := srv.Run(ctx); err != nil {
					t.Error(err)
					wg.Done()
				} else {
					go func(s *kcpServer) {
						defer wg.Done()
						if err := s.Ready(); err != nil {
							t.Errorf("kcp server %s never became ready: %v", s.name, err)
						}
					}(srv)
				}
				defer srv.GatherArtifacts()
			}
			wg.Wait()

			// if we've failed during startup, don't bother running the test
			if t.Failed() {
				return
			}
			t.Logf("Started kcp servers after %s", time.Since(start))
```

Reply:
Thanks, and what is the race you saw?

Reply:
on timeout of the server startup the test func is called. It should not be called then. I.e.
```
			if t.Failed() {
				return
			}
```

Reply:
Got it, thanks!